{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploring PMN file formats\n",
    "Specs listed [here](http://bioinformatics.ai.sri.com/ptools/flatfile-format.html#reactions.dat), playing around specifically with files from `aracyc/17.0/data/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "from os import listdir\n",
    "from os.path import join"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathbase = '../data/straying_off_topic_data/pmn_data/pmn/plantcyc/aracyc/17.0/data/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "My current thought is that, if the fields are consistent across the entire document (they may not be, there have been no promises), those fields can become column names, and each `//` `//` separated section can be a row. Will also eventually have to account for comments, but going to intentionally use a file without comments for now just to see if what I think is possible works.\n",
    "\n",
    "Update: They are definitely not consistent across the entire document (e.g. fields are just missing entirely if they are empty in a given entry). Additionally, in some files, the full list of fields is specified in the opening comment, but in some it is not. This is irritating af beacuse it means that I'll have to manually supply the possible field names from the website, but whatever. Also, everything has comments, so I'm just going to have to deal with it now.\n",
    "\n",
    "My actual question is, how on earth do they build the website from this stuff????"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also -- some files seem to have both a `.col` and a `.dat` version, do they actually contain the same information? Going to read in both for this play file and see."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading in .dat file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ah yes, some files do have the requisite ` - `.... and there are also `-`'s in the field names. *cries internally*\n",
    "\n",
    "The good news is that the attr-value separating hyphens are surrounded by spaces -- so I will look for the first occurence of ` - ` in order to separate fields from values.\n",
    "\n",
    "Fields can occur multiple times with different values... great.\n",
    "\n",
    "The docs say that for long strings the field entries can be on multiple lines... I hope to god that means thay wrap depending on the size of the screen, and not that they're actually split wiht newlines. I'm going to assume the former until proven guilty, because idk what kind of masochist would do the latter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "play_file = f'{pathbase}/genes.dat'\n",
    "play_file_fields = [\n",
    "    'UNIQUE-ID',\n",
    "    'TYPES',\n",
    "    'COMMON-NAME',\n",
    "    'CENTISOME-POSITION',\n",
    "    'CITATIONS',\n",
    "    'COMMENT',\n",
    "    'COMPONENT-OF',\n",
    "    'COMPONENTS',\n",
    "    'DBLINKS',\n",
    "    'IN-PARALOGOUS-GENE-GROUP',\n",
    "    'INTERRUPTED?',\n",
    "    'LAST-UPDATE',\n",
    "    'LEFT-END-POSITION',\n",
    "    'PRODUCT',\n",
    "    'PRODUCT-STRING',\n",
    "    'RIGHT-END-POSITION',\n",
    "    'SYNONYMS',\n",
    "    'TRANSCRIPTION-DIRECTION'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_dat(path, fields):\n",
    "    \"\"\"\n",
    "    Given a file path and a list of possible field names, create a df from a PMN .dat file.\n",
    "    \n",
    "    Edge cases accounted for:\n",
    "     - Comments\n",
    "     - Fieldnames being used multiple times with different values in the same entry\n",
    "     - Hyphens not preceeded or followed by spaces being present in fieldnames or values\n",
    "     - Empty lines\n",
    "     - // being present in URLs (separates only on // followed by a newline)\n",
    "    \n",
    "    parameters:\n",
    "        path, str: path to file to read\n",
    "        fields, list of str: list of possible field names in file to read\n",
    "        \n",
    "    returns: \n",
    "        df, pandas df: dataframe with field names as columns, with NA's in cells where the field was missing\n",
    "            for that row\n",
    "    \"\"\"\n",
    "    # Read in file\n",
    "    with open(path, encoding='windows-1250') as myf:\n",
    "        lines = myf.read()\n",
    "        \n",
    "    # Separate file on // first\n",
    "    lines = lines.split('//\\n')\n",
    "    \n",
    "    # Make a list of dicts with each dict being an entry\n",
    "    df_dicts = []\n",
    "    for entry in lines:\n",
    "        # Split on newlines to get individual entries\n",
    "        attr_vals = entry.split('\\n')\n",
    "        # Drop any empty lines\n",
    "        attr_vals = [av for av in attr_vals if av]\n",
    "        # Drop any lines beginning with #\n",
    "        attr_vals = [av for av in attr_vals if av[0][0]!='#']\n",
    "        # Split on ' - ' to get the field name and the field value in a nested list\n",
    "        attr_vals = [av.split(' - ') for av in attr_vals]\n",
    "        \n",
    "        entry_dict = {}\n",
    "        for colname in fields:\n",
    "            # Get all attr-val pairs with the fieldname\n",
    "            field_attr_vals = [av for av in attr_vals if av[0]==colname]\n",
    "            # If there's more than one entry for this fieldname, make the vals into a tuple\n",
    "            if len(field_attr_vals) > 1:\n",
    "                val = tuple([fav[1] for fav in field_attr_vals])\n",
    "            # If there aren't any entries for this fieldname, add an NA\n",
    "            elif len(field_attr_vals) == 0:\n",
    "                val = np.nan\n",
    "            # If there's only one value, add that as a single val\n",
    "            else: val = field_attr_vals[0][1]\n",
    "            # Add entry to dict\n",
    "            entry_dict[colname] = val\n",
    "            \n",
    "        df_dicts.append(entry_dict)\n",
    "\n",
    "    # Make the df\n",
    "    df = pd.DataFrame(df_dicts)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUE-ID</th>\n",
       "      <th>TYPES</th>\n",
       "      <th>COMMON-NAME</th>\n",
       "      <th>CENTISOME-POSITION</th>\n",
       "      <th>CITATIONS</th>\n",
       "      <th>COMMENT</th>\n",
       "      <th>COMPONENT-OF</th>\n",
       "      <th>COMPONENTS</th>\n",
       "      <th>DBLINKS</th>\n",
       "      <th>IN-PARALOGOUS-GENE-GROUP</th>\n",
       "      <th>INTERRUPTED?</th>\n",
       "      <th>LAST-UPDATE</th>\n",
       "      <th>LEFT-END-POSITION</th>\n",
       "      <th>PRODUCT</th>\n",
       "      <th>PRODUCT-STRING</th>\n",
       "      <th>RIGHT-END-POSITION</th>\n",
       "      <th>SYNONYMS</th>\n",
       "      <th>TRANSCRIPTION-DIRECTION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT3G25960</td>\n",
       "      <td>Unclassified-Genes</td>\n",
       "      <td>AT3G25960</td>\n",
       "      <td>40.474438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHROMOSOME-3-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(TAIR \"AT3G25960\" NIL |pzhang| 3739906193 NIL ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9499676</td>\n",
       "      <td>AT3G25960-MONOMER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9501169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G28130</td>\n",
       "      <td>Unclassified-Genes</td>\n",
       "      <td>AT4G28130</td>\n",
       "      <td>75.17636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHROMOSOME-4-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(TAIR \"AT4G28130\" NIL |pzhang| 3739906195 NIL ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13971558</td>\n",
       "      <td>AT4G28130-MONOMER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13974329</td>\n",
       "      <td>DGK6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT5G10650</td>\n",
       "      <td>ORFs</td>\n",
       "      <td>AT5G10650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(MONOMERQT-7998, MONOMERQT-7997, AT5G10650-MON...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G25055</td>\n",
       "      <td>ORFs</td>\n",
       "      <td>AT1G25055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AT1G25055-MONOMER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT3G28480</td>\n",
       "      <td>Unclassified-Genes</td>\n",
       "      <td>AT3G28480</td>\n",
       "      <td>45.49173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHROMOSOME-3-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(TAIR \"AT3G28480\" NIL |pzhang| 3739906193 NIL ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10677270</td>\n",
       "      <td>AT3G28480-MONOMER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10679625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   UNIQUE-ID               TYPES COMMON-NAME CENTISOME-POSITION CITATIONS  \\\n",
       "0  AT3G25960  Unclassified-Genes   AT3G25960      40.474438           NaN   \n",
       "1  AT4G28130  Unclassified-Genes   AT4G28130       75.17636           NaN   \n",
       "2  AT5G10650                ORFs   AT5G10650                NaN       NaN   \n",
       "3  AT1G25055                ORFs   AT1G25055                NaN       NaN   \n",
       "4  AT3G28480  Unclassified-Genes   AT3G28480       45.49173           NaN   \n",
       "\n",
       "  COMMENT    COMPONENT-OF  COMPONENTS  \\\n",
       "0     NaN  CHROMOSOME-3-4         NaN   \n",
       "1     NaN  CHROMOSOME-4-7         NaN   \n",
       "2     NaN             NaN         NaN   \n",
       "3     NaN             NaN         NaN   \n",
       "4     NaN  CHROMOSOME-3-5         NaN   \n",
       "\n",
       "                                             DBLINKS  \\\n",
       "0  (TAIR \"AT3G25960\" NIL |pzhang| 3739906193 NIL ...   \n",
       "1  (TAIR \"AT4G28130\" NIL |pzhang| 3739906195 NIL ...   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  (TAIR \"AT3G28480\" NIL |pzhang| 3739906193 NIL ...   \n",
       "\n",
       "   IN-PARALOGOUS-GENE-GROUP  INTERRUPTED?  LAST-UPDATE LEFT-END-POSITION  \\\n",
       "0                       NaN           NaN          NaN           9499676   \n",
       "1                       NaN           NaN          NaN          13971558   \n",
       "2                       NaN           NaN          NaN               NaN   \n",
       "3                       NaN           NaN          NaN               NaN   \n",
       "4                       NaN           NaN          NaN          10677270   \n",
       "\n",
       "                                             PRODUCT  PRODUCT-STRING  \\\n",
       "0                                  AT3G25960-MONOMER             NaN   \n",
       "1                                  AT4G28130-MONOMER             NaN   \n",
       "2  (MONOMERQT-7998, MONOMERQT-7997, AT5G10650-MON...             NaN   \n",
       "3                                  AT1G25055-MONOMER             NaN   \n",
       "4                                  AT3G28480-MONOMER             NaN   \n",
       "\n",
       "  RIGHT-END-POSITION SYNONYMS TRANSCRIPTION-DIRECTION  \n",
       "0            9501169      NaN                     NaN  \n",
       "1           13974329     DGK6                     NaN  \n",
       "2                NaN      NaN                     NaN  \n",
       "3                NaN      NaN                     NaN  \n",
       "4           10679625      NaN                     NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "play_df = read_dat(play_file, play_file_fields)\n",
    "play_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read in the corresponding .col file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There's a bigass comment section at the top of this too, ffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUE-ID</th>\n",
       "      <th>NAME</th>\n",
       "      <th>PRODUCT-NAME</th>\n",
       "      <th>SWISS-PROT-ID</th>\n",
       "      <th>REPLICON</th>\n",
       "      <th>START-BASE</th>\n",
       "      <th>END-BASE</th>\n",
       "      <th>SYNONYMS</th>\n",
       "      <th>SYNONYMS.1</th>\n",
       "      <th>SYNONYMS.2</th>\n",
       "      <th>SYNONYMS.3</th>\n",
       "      <th>GENE-CLASS</th>\n",
       "      <th>GENE-CLASS.1</th>\n",
       "      <th>GENE-CLASS.2</th>\n",
       "      <th>GENE-CLASS.3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT3G25960</td>\n",
       "      <td>AT3G25960</td>\n",
       "      <td>pyruvate kinase</td>\n",
       "      <td>Q9LU95</td>\n",
       "      <td>CHROMOSOME-3</td>\n",
       "      <td>9499676.0</td>\n",
       "      <td>9501169.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G28130</td>\n",
       "      <td>AT4G28130</td>\n",
       "      <td>diacylglycerol kinase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHROMOSOME-4</td>\n",
       "      <td>13971558.0</td>\n",
       "      <td>13974329.0</td>\n",
       "      <td>DGK6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT5G10650</td>\n",
       "      <td>AT5G10650</td>\n",
       "      <td>AT5G10650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G25055</td>\n",
       "      <td>AT1G25055</td>\n",
       "      <td>AT1G25055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT3G28480</td>\n",
       "      <td>AT3G28480</td>\n",
       "      <td>oxidoreductase, acting on paired donors, with ...</td>\n",
       "      <td>Q9LSI6</td>\n",
       "      <td>CHROMOSOME-3</td>\n",
       "      <td>10677270.0</td>\n",
       "      <td>10679625.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   UNIQUE-ID       NAME                                       PRODUCT-NAME  \\\n",
       "0  AT3G25960  AT3G25960                                    pyruvate kinase   \n",
       "1  AT4G28130  AT4G28130                              diacylglycerol kinase   \n",
       "2  AT5G10650  AT5G10650                                          AT5G10650   \n",
       "3  AT1G25055  AT1G25055                                          AT1G25055   \n",
       "4  AT3G28480  AT3G28480  oxidoreductase, acting on paired donors, with ...   \n",
       "\n",
       "  SWISS-PROT-ID      REPLICON  START-BASE    END-BASE SYNONYMS SYNONYMS.1  \\\n",
       "0        Q9LU95  CHROMOSOME-3   9499676.0   9501169.0      NaN        NaN   \n",
       "1           NaN  CHROMOSOME-4  13971558.0  13974329.0     DGK6        NaN   \n",
       "2           NaN           NaN         NaN         NaN      NaN        NaN   \n",
       "3           NaN           NaN         NaN         NaN      NaN        NaN   \n",
       "4        Q9LSI6  CHROMOSOME-3  10677270.0  10679625.0      NaN        NaN   \n",
       "\n",
       "  SYNONYMS.2 SYNONYMS.3    GENE-CLASS  GENE-CLASS.1  GENE-CLASS.2  \\\n",
       "0        NaN        NaN  UNCLASSIFIED           NaN           NaN   \n",
       "1        NaN        NaN  UNCLASSIFIED           NaN           NaN   \n",
       "2        NaN        NaN           NaN           NaN           NaN   \n",
       "3        NaN        NaN           NaN           NaN           NaN   \n",
       "4        NaN        NaN  UNCLASSIFIED           NaN           NaN   \n",
       "\n",
       "   GENE-CLASS.3  \n",
       "0           NaN  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "play_col_path = f'{pathbase}/genes.col'\n",
    "play_col = pd.read_csv(play_col_path, sep='\\t', comment='#')\n",
    "play_col.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if the unique-name cols have the same vals\n",
    "dat_genes = set(play_df['UNIQUE-ID'])\n",
    "col_genes = set(play_col['UNIQUE-ID'])\n",
    "dat_genes == col_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6346 6345 6345\n"
     ]
    }
   ],
   "source": [
    "overlap = dat_genes.intersection(col_genes)\n",
    "print(len(dat_genes), len(col_genes), len(overlap))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like these files encode totally different information about more or less the same genes... There is one gene missing from the col file that's present in the dat file. WHY ARE THERE TWO DIFFERENT FILES FOR THIS IM SCREAMING"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Deciding what data can be useful\n",
    "Want to choose file types and categories that I'll actually utilize in my relation extraction pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biopax-level2.owl\t\tgenes.dat\t\t pubs.dat\r\n",
      "biopax-level3.owl\t\tmetabolic-reactions.xml  reaction-links.dat\r\n",
      "classes.dat\t\t\toverview.graph\t\t reactions.dat\r\n",
      "compound-links.dat\t\tpathway-links.dat\t regulation.dat\r\n",
      "compounds.dat\t\t\tpathways.col\t\t regulons.dat\r\n",
      "dnabindsites.dat\t\tpathways.dat\t\t rnas.dat\r\n",
      "enzrxns.dat\t\t\tpromoters.dat\t\t species.dat\r\n",
      "enzymes.col\t\t\tprotcplxs.col\t\t terminators.dat\r\n",
      "gene_association.aracyc\t\tprotein-features.dat\t transporters.col\r\n",
      "gene_association.aracyc-errors\tprotein-links.dat\t transunits.dat\r\n",
      "gene-links.dat\t\t\tproteins.dat\r\n",
      "genes.col\t\t\tprotligandcplxes.dat\r\n"
     ]
    }
   ],
   "source": [
    "!ls ../data/straying_off_topic_data/pmn_data/pmn/plantcyc/aracyc/17.0/data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Araport11_genes.201606.pep.repr.fasta.pmn.e2p2v3.orxn.pf\r\n",
      "at_chloroplast.pf\r\n",
      "at_chrom1.pf\r\n",
      "at_chrom2.pf\r\n",
      "at_chrom3.pf\r\n",
      "at_chrom4.pf\r\n",
      "at_chrom5.pf\r\n",
      "at_genetic_loci.pf\r\n",
      "at_mito.pf\r\n",
      "genetic-elements.dat\r\n",
      "organism.dat\r\n",
      "organism.dat~\r\n",
      "organism-init.dat\r\n",
      "sample-genetic-elements.dat\r\n"
     ]
    }
   ],
   "source": [
    "!ls ../data/straying_off_topic_data/pmn_data/pmn/plantcyc/aracyc/17.0/input/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AraCyc</td>\n",
       "      <td>AT3G09260-MONOMER</td>\n",
       "      <td>AT3G09260</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0010168</td>\n",
       "      <td>PMID:12581307</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>scopolin beta-glucosidase</td>\n",
       "      <td>AT3G09260|PYK10|PSR3.1|LEB|BGLU23|AT3G09260</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:</td>\n",
       "      <td>20130130</td>\n",
       "      <td>AraCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AraCyc</td>\n",
       "      <td>AT3G09260-MONOMER</td>\n",
       "      <td>AT3G09260</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005783</td>\n",
       "      <td>PMID:22923678</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>scopolin beta-glucosidase</td>\n",
       "      <td>AT3G09260|PYK10|PSR3.1|LEB|BGLU23|AT3G09260</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:</td>\n",
       "      <td>20130130</td>\n",
       "      <td>AraCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AraCyc</td>\n",
       "      <td>AT5G48950-MONOMER</td>\n",
       "      <td>AT5G48950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005777</td>\n",
       "      <td>PMID:22372525</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>1,4-dihydroxy-2-naphthoyl-CoA thioesterase 2</td>\n",
       "      <td>DHNAT2|AT5G48950</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:</td>\n",
       "      <td>20120420</td>\n",
       "      <td>AraCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AraCyc</td>\n",
       "      <td>AT2G40690-MONOMER</td>\n",
       "      <td>AT2G40690</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0031969</td>\n",
       "      <td>PMID:20061580</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>glycerol 3-phosphate dehydrogenase</td>\n",
       "      <td>SFD1|GLY1|AT2G40690</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:</td>\n",
       "      <td>20120402</td>\n",
       "      <td>AraCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AraCyc</td>\n",
       "      <td>AT2G30140-MONOMER</td>\n",
       "      <td>AT2G30140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>PMID:22404750</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>UDP-glycosyltransferase 87A2</td>\n",
       "      <td>UGT87A2|AT2G30140</td>\n",
       "      <td>protein</td>\n",
       "      <td>taxon:</td>\n",
       "      <td>20120426</td>\n",
       "      <td>AraCyc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0                  1          2   3           4              5    6   \\\n",
       "0  AraCyc  AT3G09260-MONOMER  AT3G09260 NaN  GO:0010168  PMID:12581307  IDA   \n",
       "1  AraCyc  AT3G09260-MONOMER  AT3G09260 NaN  GO:0005783  PMID:22923678  IDA   \n",
       "2  AraCyc  AT5G48950-MONOMER  AT5G48950 NaN  GO:0005777  PMID:22372525  IDA   \n",
       "3  AraCyc  AT2G40690-MONOMER  AT2G40690 NaN  GO:0031969  PMID:20061580  IDA   \n",
       "4  AraCyc  AT2G30140-MONOMER  AT2G30140 NaN  GO:0005737  PMID:22404750  IDA   \n",
       "\n",
       "   7  8                                             9   \\\n",
       "0 NaN  C                     scopolin beta-glucosidase   \n",
       "1 NaN  C                     scopolin beta-glucosidase   \n",
       "2 NaN  C  1,4-dihydroxy-2-naphthoyl-CoA thioesterase 2   \n",
       "3 NaN  C            glycerol 3-phosphate dehydrogenase   \n",
       "4 NaN  C                  UDP-glycosyltransferase 87A2   \n",
       "\n",
       "                                            10       11      12        13  \\\n",
       "0  AT3G09260|PYK10|PSR3.1|LEB|BGLU23|AT3G09260  protein  taxon:  20130130   \n",
       "1  AT3G09260|PYK10|PSR3.1|LEB|BGLU23|AT3G09260  protein  taxon:  20130130   \n",
       "2                             DHNAT2|AT5G48950  protein  taxon:  20120420   \n",
       "3                          SFD1|GLY1|AT2G40690  protein  taxon:  20120402   \n",
       "4                            UGT87A2|AT2G30140  protein  taxon:  20120426   \n",
       "\n",
       "       14  \n",
       "0  AraCyc  \n",
       "1  AraCyc  \n",
       "2  AraCyc  \n",
       "3  AraCyc  \n",
       "4  AraCyc  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aracyc = pd.read_csv('../data/straying_off_topic_data/pmn_data/pmn/plantcyc/aracyc/17.0/data/gene_association.aracyc', sep='\\t', header=None)\n",
    "aracyc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thank you for having no information whatsoever about wtf the contents of this file are............"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To save myself a little heartache in the future, I think I'm going to make a file that contains the field names for the various types of files listed on the specs page, it'll be a json where the key is the file name and the values are lists of the field names. This will also help me see what kind of data each one has & whether or not I want to use it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "field_json = {\n",
    "    'classes.dat':['UNIQUE-ID','TYPES','COMMENT','COMMON-NAME','SYNONYMS'],\n",
    "    'bindrxns.dat':['UNIQUE-ID','TYPES','ACTIVATORS','INHIBITORS','OFFICIAL-EC','REACTANTS'],\n",
    "    'compounds.dat':['UNIQUE-ID','TYPES','COMMON-NAME','ABBREV-NAME','ATOM-CHARGES','CHEMICAL-FORMULA',\n",
    "                    'CITATIONS','COFACTORS-OF','COFACTORS-OR-PROSTHETIC-GROUPS-OF','COMMENT','COMPONENT-OF',\n",
    "                    'CREDITS','DATA-SOURCE','DBLINKS','INCHI','MOLECULAR-WEIGHT','N+1-NAME','N-1-NAME','N-NAME',\n",
    "                    'PKA1','PKA2','PKA3','PROSTHETIC-GROUPS-OF','REGULATES','SMILES','SUPERATOMS','SYNONYMS',\n",
    "                    'SYSTEMATIC-NAME'],\n",
    "    'dnabindsites.dat':['UNIQUE-ID','TYPES','ABS-CENTER-POS','CITATIONS','COMMENT','COMPONENT-OF','DBLINKS',\n",
    "                        'INVOLVED-IN-REGULATION','REGULATED-PROMOTER','RELATIVE-CENTER-DISTANCE','SYNONYMS',\n",
    "                        'TYPE-OF-EVIDENCE'],\n",
    "    'enzrxns.dat':['UNIQUE-ID','TYPES','COMMON-NAME','ALTERNATIVE-COFACTORS','ALTERNATIVE-SUBSTRATES','CITATIONS',\n",
    "                    'COFACTOR-BINDING-COMMENT','COFACTORS','COFACTORS-OR-PROSTHETIC-GROUPS','COMMENT','ENZYME','KM',\n",
    "                    'PH-OPT','PROSTHETIC-GROUPS','REACTION','REACTION-DIRECTION','REGULATED-BY',\n",
    "                    'REQUIRED-PROTEIN-COMPLEX','SYNONYMS','TEMPERATURE-OPT'],\n",
    "    'genes.dat':['UNIQUE-ID','TYPES','COMMON-NAME','CENTISOME-POSITION','CITATIONS','COMMENT','COMPONENT-OF',\n",
    "                'COMPONENTS','DBLINKS','IN-PARALOGOUS-GENE-GROUP','INTERRUPTED?','LAST-UPDATE','LEFT-END-POSITION',\n",
    "                'PRODUCT','PRODUCT-STRING','RIGHT-END-POSITION','SYNONYMS','TRANSCRIPTION-DIRECTION'],\n",
    "    'pathways.dat':['UNIQUE-ID','TYPES','COMMON-NAME','CITATIONS','CLASS-INSTANCE-LINKS','COMMENT','CREDITS',\n",
    "                    'DBLINKS','ENZYME-USE','HYPOTHETICAL-REACTIONS','IN-PATHWAY','NET-REACTION-EQUATION',\n",
    "                    'PATHWAY-INTERACTIONS','PATHWAY-LINKS','POLYMERIZATION-LINKS','PREDECESSORS','PRIMARIES',\n",
    "                    'REACTION-LAYOUT','REACTION-LIST','SPECIES','SUB-PATHWAYS','SUPER-PATHWAYS','SYNONYMS'],\n",
    "    'promoters.dat':['UNIQUE-ID','TYPES','COMMON-NAME','ABSOLUTE-PLUS-1-POS','BINDS-SIGMA-FACTOR','CITATIONS',\n",
    "                    'COMMENT','COMPONENT-OF','DBLINKS','PROMOTER-EVIDENCE','REGULATED-BY','SYNONYMS'],\n",
    "    'protein-features.dat':['UNIQUE-ID','TYPES','COMMON-NAME','COMMENT','ATTACHED-GROUP','CITATIONS','FEATURE-OF',\n",
    "                            'HOMOLOGY-MOTIF','LEFT-END-POSITION','POSSIBLE-FEATURE-STATES','RESIDUE-NUMBER',\n",
    "                            'RESIDUE-TYPE','RIGHT-END-POSITION'],\n",
    "    'proteins.dat':['UNIQUE-ID','TYPES','COMMON-NAME','CATALYZES','CHEMICAL-FORMULA','CITATIONS','COMMENT',\n",
    "                    'COMPONENT-OF','CREDITS','DBLINKS','DNA-FOOTPRINT-SIZE','FEATURES','GENE','GO-TERMS',\n",
    "                    'ISOZYME-SEQUENCE-SIMILARITY','LOCATIONS','MODIFIED-FORM','MOLECULAR-WEIGHT-EXP',\n",
    "                    'MOLECULAR-WEIGHT-KD','MOLECULAR-WEIGHT-SEQ','PI','REGULATES','SPECIES','SPLICE-FORM-INTRONS',\n",
    "                    'SYMMETRY','SYNONYMS','UNMODIFIED-FORM'],\n",
    "    'protligandcplxs.dat':['UNIQUE-ID','TYPES','CATALYZES','COMMON-NAME','COMMENT','COMPONENTS','DBLINKS',\n",
    "                           'DNA-FOOTPRINT-SIZE','MOLECULAR-WEIGHT-KD','MOLECULAR-WEIGHT-SEQ','REGULATES','SYMMETRY',\n",
    "                           'SYNONYMS'],\n",
    "    'pubs.dat':['UNIQUE-ID','TYPES','ABSTRACT','AUTHORS','COMMENT','MEDLINE-UID','PUBMED-ID','REFERENT-FRAME',\n",
    "                'SOURCE','TITLE','YEAR'],\n",
    "    'reactions.dat':['UNIQUE-ID','TYPES','COMMON-NAME','CITATIONS','COMMENT','DELTAG0','EC-NUMBER',\n",
    "                     'ENZYMATIC-REACTION','IN-PATHWAY','LEFT','OFFICIAL-EC?','ORPHAN?','RIGHT','SIGNAL','SPECIES',\n",
    "                     'SPONTANEOUS?','SYNONYMS'],\n",
    "    'regulation.dat':['UNIQUE-ID','TYPES','COMMON-NAME','ASSOCIATED-BINDING-SITE','COMMENT','MECHANISM','MODE',\n",
    "                      'PHYSIOLOGICALLY-RELEVANT?','REGULATED-BY','REGULATED-ENTITY','REGULATOR','SYNONYMS'],\n",
    "    'regulons.dat':['UNIQUE-ID','TYPES','COMMON-NAME','ACTIVATORS-ALLOSTERIC-OF','ACTIVATORS-NONALLOSTERIC-OF',\n",
    "                    'ACTIVATORS-UNKMECH-OF','APPEARS-IN-BINDING-REACTIONS','AROMATIC-RINGS','ATOM-CHARGES',\n",
    "                    'ATOM-CHIRALITY','CATALYZES','CHARGE','CHEMICAL-FORMULA','CITATIONS','COFACTORS-OF',\n",
    "                    'COFACTORS-OR-PROSTHETIC-GROUPS-OF','COMMENT','COMPONENT-COEFFICIENTS','COMPONENT-OF','COMPONENTS',\n",
    "                    'CREDITS','DATA-SOURCE','DBLINKS','DNA-FOOTPRINT-SIZE','FEATURES','FUNCTIONAL-ASSIGNMENT-COMMENT',\n",
    "                    'FUNCTIONAL-ASSIGNMENT-STATUS','GENE','GO-TERMS','INHIBITORS-ALLOSTERIC-OF',\n",
    "                    'INHIBITORS-COMPETITIVE-OF','INHIBITORS-IRREVERSIBLE-OF','INHIBITORS-NONCOMPETITIVE-OF',\n",
    "                    'INHIBITORS-OTHER-OF','INHIBITORS-UNCOMPETITIVE-OF','INHIBITORS-UNKMECH-OF',\n",
    "                    'INSTANCE-NAME-TEMPLATE','ISOZYME-SEQUENCE-SIMILARITY','LOCATIONS','MODIFIED-FORM',\n",
    "                    'MOLECULAR-WEIGHT','MOLECULAR-WEIGHT-EXP','MOLECULAR-WEIGHT-KD','MOLECULAR-WEIGHT-SEQ','N+1-NAME',\n",
    "                    'N-1-NAME','N-NAME','NEIDHARDT-SPOT-NUMBER','PI','PROSTHETIC-GROUPS-OF','REGULATED-BY','REGULATES',\n",
    "                    'SPECIES','SPLICE-FORM-INTRONS','STRUCTURE-BONDS','SYMMETRY','SYNONYMS','UNMODIFIED-FORM'],\n",
    "    'terminators.dat':['UNIQUE-ID','TYPES','COMMON-NAME','APPEARS-IN-BINDING-REACTIONS','CITATIONS','COMMENT',\n",
    "                       'COMPONENT-OF','COMPONENTS','CREDITS','DATA-SOURCE','DBLINKS','INSTANCE-NAME-TEMPLATE',\n",
    "                       'LEFT-END-POSITION','RIGHT-END-POSITION','SYNONYMS'],\n",
    "    'transunits.dat':['UNIQUE-ID','TYPES','COMMON-NAME','APPEARS-IN-BINDING-REACTIONS','CITATIONS','COMMENT',\n",
    "                      'COMPONENT-OF','COMPONENTS','CREDITS','DATA-SOURCE','DBLINKS','EXTENT-UNKNOWN?',\n",
    "                      'INSTANCE-NAME-TEMPLATE','LEFT-END-POSITION','REGULATED-BY','RIGHT-END-POSITION','SYNONYMS']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../models/benchmarks/pmn_relation_matching/pmn_dat_fields.json', 'w') as myf:\n",
    "    json.dump(field_json, myf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Please, thank me for my service."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Questions that arose during making this:\n",
    "1. Is UNIQUE-ID always a gene name?\n",
    "\n",
    "Ok well that was basically the one question I guess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's just read 'em in lads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading in file pubs.dat\n",
      "Success!\n",
      "\n",
      "Reading in file dnabindsites.dat\n",
      "Success!\n",
      "\n",
      "Reading in file pathways.dat\n",
      "Success!\n",
      "\n",
      "Reading in file classes.dat\n",
      "Success!\n",
      "\n",
      "Reading in file protein-features.dat\n",
      "Success!\n",
      "\n",
      "Reading in file regulation.dat\n",
      "Success!\n",
      "\n",
      "Reading in file compounds.dat\n",
      "Success!\n",
      "\n",
      "Reading in file reactions.dat\n",
      "Success!\n",
      "\n",
      "Reading in file terminators.dat\n",
      "Success!\n",
      "\n",
      "Reading in file genes.dat\n",
      "Success!\n",
      "\n",
      "Reading in file promoters.dat\n",
      "Success!\n",
      "\n",
      "Reading in file enzrxns.dat\n",
      "Success!\n",
      "\n",
      "Reading in file transunits.dat\n",
      "Success!\n",
      "\n",
      "Reading in file proteins.dat\n",
      "Success!\n",
      "\n",
      "Reading in file regulons.dat\n",
      "Success!\n"
     ]
    }
   ],
   "source": [
    "data_files = []\n",
    "for f in listdir(pathbase):\n",
    "    if f in field_json.keys():\n",
    "        print(f'\\nReading in file {f}')\n",
    "        try:\n",
    "            df = read_dat(join(pathbase, f), field_json[f])\n",
    "            data_files.append(df)\n",
    "            print('Success!')\n",
    "        except UnicodeDecodeError as e:\n",
    "            print(f'Cannot read in {f}, error is:\\n{e}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see if some of these files are encoded with a different character set, or if something is just broken:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -orch (/opt/software/Python/3.6.4-foss-2018a/lib/python3.6/site-packages)\u001b[0m\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -orch (/opt/software/Python/3.6.4-foss-2018a/lib/python3.6/site-packages)\u001b[0m\n",
      "Collecting charset-normalizer\n",
      "  Using cached charset_normalizer-2.0.12-py3-none-any.whl (39 kB)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -orch (/opt/software/Python/3.6.4-foss-2018a/lib/python3.6/site-packages)\u001b[0m\n",
      "Installing collected packages: charset-normalizer\n",
      "\u001b[33m  WARNING: The script normalizer is installed in '/mnt/home/lotrecks/.local/bin' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\n",
      "Successfully installed charset-normalizer-2.0.12\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -orch (/opt/software/Python/3.6.4-foss-2018a/lib/python3.6/site-packages)\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install charset-normalizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "from charset_normalizer import detect\n",
    "import urllib.request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_encodings = {}\n",
    "for f in listdir(pathbase):\n",
    "    if f in field_json.keys():\n",
    "        with open(join(pathbase, f), 'rb') as myf:\n",
    "            rawdata = myf.read()\n",
    "        file_encodings[f] = detect(rawdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pubs.dat': {'encoding': 'windows-1250',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'dnabindsites.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'pathways.dat': {'encoding': 'windows-1250',\n",
       "  'language': 'English',\n",
       "  'confidence': 0.9482},\n",
       " 'classes.dat': {'encoding': 'windows-1250',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'protein-features.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'regulation.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'compounds.dat': {'encoding': 'windows-1250',\n",
       "  'language': 'English',\n",
       "  'confidence': 0.9386},\n",
       " 'reactions.dat': {'encoding': 'windows-1250',\n",
       "  'language': 'Indonesian',\n",
       "  'confidence': 0.9104},\n",
       " 'terminators.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'genes.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 0.9806},\n",
       " 'promoters.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'enzrxns.dat': {'encoding': 'windows-1250',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'transunits.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0},\n",
       " 'proteins.dat': {'encoding': 'windows-1250',\n",
       "  'language': 'English',\n",
       "  'confidence': 0.9638},\n",
       " 'regulons.dat': {'encoding': 'ascii',\n",
       "  'language': 'English',\n",
       "  'confidence': 1.0}}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_encodings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I tried reading every file in with both `ascii` and `windows-1250` to see if Galen was right about them just not tripping errors for utf-8 on some of them even though they're technically encoded differently, and `windows-1250` works!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dygiepp",
   "language": "python",
   "name": "dygiepp"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
